crates/wast/src/ast/component.rs

```diff
@@ -0,0 +1,188 @@
+    ///
+    /// # Errors
+    ///
+    /// If an error happens during resolution, such a name resolution error or
+    /// items are found in the wrong order, then an error is returned.
+    pub fn resolve(&mut self) -> std::result::Result<(), crate::Error> {
+        Ok(())
+    }
+
+    /// Encodes this [`Component`] to its binary form.
+    ///
+    /// This function will take the textual representation in [`Component`] and
+    /// perform all steps necessary to convert it to a binary WebAssembly
+    /// component, suitable for writing to a `*.wasm` file. This function may
+    /// internally modify the [`Component`], for example:
+    ///
+    /// * Name resolution is performed to ensure that `Index::Id` isn't present
+    ///   anywhere in the AST.
+    ///
+    /// * Inline shorthands such as imports/exports/types are all expanded to be
+    ///   dedicated fields of the component.
+    ///
+    /// * Component fields may be shuffled around to preserve index ordering from
+    ///   expansions.
+    ///
+    /// After all of this expansion has happened the component will be converted to
+    /// its binary form and returned as a `Vec<u8>`. This is then suitable to
+    /// hand off to other wasm runtimes and such.
+    ///
+    /// # Errors
+    ///
+    /// This function can return an error for name resolution errors and other
+    /// expansion-related errors.
+    pub fn encode(&mut self) -> std::result::Result<Vec<u8>, crate::Error> {
+        self.resolve()?;
+        Ok(crate::binary::encode_component(self))
+    }
+
+    pub(super) fn validate(&self, parser: Parser<'_>) -> Result<()> {
+        let mut starts = 0;
+        if let ComponentKind::Text(fields) = &self.kind {
+            for item in fields.iter() {
+                if let ComponentField::Start(_) = item {
+                    starts += 1;
+                }
+            }
+        }
+        if starts > 1 {
+            return Err(parser.error("multiple start sections found"));
+        }
+        Ok(())
+    }
+}
+
+impl<'a> Parse<'a> for Component<'a> {
+    fn parse(parser: Parser<'a>) -> Result<Self> {
+        let _r = parser.register_annotation("custom");
+
+        let span = parser.parse::<kw::component>()?.0;
+        let id = parser.parse()?;
+        let name = parser.parse()?;
+
+        let kind = if parser.peek::<kw::binary>() {
+            parser.parse::<kw::binary>()?;
+            let mut data = Vec::new();
+            while !parser.is_empty() {
+                data.push(parser.parse()?);
+            }
+            ComponentKind::Binary(data)
+        } else {
+            ComponentKind::Text(ComponentField::parse_remaining(parser)?)
+        };
+        Ok(Component {
+            span,
+            id,
+            name,
+            kind,
+        })
+    }
+}
+
+/// A listing of all possible fields that can make up a WebAssembly component.
+#[allow(missing_docs)]
+#[derive(Debug)]
+pub enum ComponentField<'a> {
+    Type(ast::Type<'a>),
```

Comment:
I think that this'll want to be `ComponentType<'a>` or something like that in the near future, and then the module/instance types would come out of `ast::Type`

I think this actually applies to most of the fields below, given the disjoint-ness of components and modules right now. There probably won't be all that much that's shared between the two parsing-wise.